### PR TITLE
fix: handling of variable and function declarations inside autoqasm programs

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*
- [Passing arguments directly to function call doesn't give the generated variable a name](https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=32619483)
- [Nested recursive functions raise "no autograph_artifact attribute" error](https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=33517696)

*Description of changes:*
- Imports `autograph_artifact` into the `operators` module to fix an error when defining a Python function inside an `aq.function`.
- Defines `aq.ArrayVar`, `aq.BitVar`, `aq.BoolVar`, `aq.FloatVar`, `aq.IntVar` as subclasses of the respective `oqpy` classes, but initialized with deterministically-generated names.
- Consolidate all usage of `next_var_name` into the newly defined types above.

*Testing done:*
- New tests to verify fixed scenarios.
- Updated all other tests to pass (some variable names changed due to the enforcement of deterministic names for all `aq` variables, even those that don't end up being declared in the program).

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/experimental/autoqasm/api.py [hunk 1]
```diff
@@ -335,7 +334,7 @@ def _make_return_instance(
 
     return_instance = None
     if return_type and aq_types.is_qasm_type(return_type):
-        return_instance = return_type(name=program_conversion_context.next_var_name(return_type))
+        return_instance = return_type()
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/api.py:337]: Cool!! I like this. I love when bugfixes point out cleaner code

File: src/braket/experimental/autoqasm/operators/__init__.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 # We need to either implement these, or determine they are not needed and remove them.
 
 # Operators below are imported directly from core autograph implementation
+from braket.experimental.autoqasm.autograph.impl.api_core import autograph_artifact  # noqa: F401
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/operators/__init__.py:24]: What a weird bug

File: src/braket/experimental/autoqasm/operators/conditional_expressions.py [hunk 3]
```diff
@@ -51,7 +51,6 @@ def _oqpy_if_exp(
     """Overload of if_exp that stages an oqpy conditional."""
     oqpy_program = program.get_program_conversion_context().get_oqpy_program()
     if isinstance(cond, oqpy.base.Var) and cond.name not in oqpy_program.declared_vars.keys():
-        cond.name = program.get_program_conversion_context().next_var_name(type(cond))
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/operators/conditional_expressions.py:54]: :tada:

File: test/unit_tests/braket/experimental/autoqasm/test_api.py [hunk 4]
```diff
@@ -792,3 +792,26 @@ def main() -> None:
...
+        def ghz(n: int):
+            if n == 1:
+                h(0)
+            else:
+                ghz(n - 1)
+                cnot(0, n - 1)
+
+        ghz(n)
+
+    expected = """OPENQASM 3.0;
```
Comment (laurencap) [inline test/unit_tests/braket/experimental/autoqasm/test_api.py:809]: cool test

**Review by laurencap — APPROVED** [review]
These changes are beautiful, I wish I had thought of writing it this way from the start!!

File: test/unit_tests/braket/experimental/autoqasm/test_api.py (hunk 4 above)
Comment (rmshaffer) [reply test/unit_tests/braket/experimental/autoqasm/test_api.py:809]: You can thank @godott for this test (which uncovered the bug), it's copied directly from his bug report.

File: src/braket/experimental/autoqasm/operators/__init__.py (hunk 2 above)
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/operators/__init__.py:24]: Yeah, this was just one of these weird autograph implementation details that I missed. At least the fix is easy.

**Pull request merged**